cirq-core/cirq/protocols/measurement_key_protocol.py

```diff
@@ -50,18 +51,36 @@ class SupportsMeasurementKey(Protocol):
     def _is_measurement_(self) -> bool:
         """Return if this object is (or contains) a measurement."""
 
+    @doc_private
+    def _measurement_key_obj_(self) -> value.MeasurementKey:
```

Comment:
Should the class docstring also mention implementing these methods? Can a class fulfill the `SupportsMeasurementKey` protocol while implementing only one of these methods? _Should_ classes ever implement both methods?

Reply:
Note that Cirq has [`@value_alternative`](https://github.com/quantumlib/Cirq/blob/9f371b0d57ff27b2df4c823f13b9c81444e64a7c/cirq-core/cirq/value/abc_alt.py#L33) for enabling users to implement a protocol with only a subset of the methods it defines. For example:
https://github.com/quantumlib/Cirq/blob/9f371b0d57ff27b2df4c823f13b9c81444e64a7c/cirq-core/cirq/devices/noise_model.py#L112-L116
This will auto-define `noisy_moment` and `noisy_operation` in terms of `noisy_moments` if it is the only method an implementation provides.

Reply:
Good point. I guess this is a broader question that applies to even the original protocol. A gate can implement both `_measurement_key_` and `_measurement_keys_` but we prioritize `keys` above `key`. Even `protocols.measurement_key` will first look up `_measurement_keys_` and only fall back to `_measurement_key_` if needed. Same applies for other flavors in this protocol set. This can lead to some inconsistent and counter-intuitive behaviors. I think we should enforce that only one of these is implemented for any concrete `Gate` or `Operation`. But structures like `Moment` and `Circuit` should be able to implement them all and delegate to constituent ops.

Maybe we can discuss this in detail and fix it in a subsequent PR. How about I just document the prioritization here for now?

Reply:
Documenting for now and fixing in a follow-up sounds good to me.